### PR TITLE
Add liveness/readiness probe parameters

### PR DESCRIPTION
**Pull request by timstclair**
- PeriodSeconds - How often to probe
- SuccessThreshold - Number of successful probes to go from failure to success state
- FailureThreshold - Number of failing probes to go from success to failure state

This commit includes to changes in behavior:
1. InitialDelaySeconds now defaults to 10 seconds, rather than the
   kubelet sync interval (although that also defaults to 10 seconds).
2. Prober only retries on probe error, not failure. To compensate, the
   default FailureThreshold is set to the maxRetries, 3.

Fixes https://github.com/kubernetes/kubernetes/issues/12866

cc/ @yujuhong

### CODE REVIEW

**Comment by timstclair** [comment]
Rebased

**Comment by timstclair** [comment]
rebased

**Comment by yujuhong** [comment]
LGTM overall. Needs a rebase again.

cc @bgrant0607 for an lgm on the changes in the api.

**Comment by timstclair** [comment]
Rebased.

File: pkg/api/v1/types.go [hunk 1]
```diff
@@ -846,17 +846,26 @@ type ExecAction struct {
 	Command []string `json:"command,omitempty"`
 }
 
-// Probe describes a liveness probe to be examined to the container.
+// Probe describes a health check probe to be examined to the container.
```
Comment (smarterclayton) [inline pkg/api/v1/types.go]: This description needs to be reworded:

`Probe describes a check to be performed against a container to determine whether it is alive or ready to take traffic.`

**Comment by dchen1107** [comment]
LGTM except above nit brought up by @smarterclayton 

Sorry for delaying this. Try to avoid rebase conflicts. ;-P

File: pkg/api/v1/types.go (hunk 1 above)
Comment (timstclair) [inline pkg/api/v1/types.go]: Thanks, done.

**Comment by dchen1107** [comment]
LGTM. please squash, then we are ready to go. Thanks!

**Comment by timstclair** [comment]
Squashed. Thanks!

**Comment by dchen1107** [comment]
@timstclair Please run verify-generated-swagger-docs.sh to fix out-of-date doc.

17:18:50 +++ /workspace/kubernetes/_tmp/pkg/api/v1/types_swagger_doc_generated.go	2015-10-31 00:14:31.787411517 +0000
17:18:50 @@ -1068,7 +1068,7 @@ func (PodTemplateSpec) SwaggerDoc() map[
17:18:50  }
17:18:50  
17:18:50  var map_Probe = map[string]string{
17:18:50 -	"": "Probe describes a health check to be performed against a container to determine whether it is alive or ready to recieve traffic.",
17:18:50 +	"": "Probe describes a health check probe to be examined to the container.",
17:18:50  	"initialDelaySeconds": "Number of seconds after the container has started before liveness probes are initiated. More info: http://releases.k8s.io/HEAD/docs/user-guide/pod-states.md#container-probes",
17:18:50  	"timeoutSeconds":      "Number of seconds after which the probe times out. Defaults to 1 second. More info: http://releases.k8s.io/HEAD/docs/user-guide/pod-states.md#container-probes",
17:18:50  	"periodSeconds":       "How often (in seconds) to perform the probe. Default to 10 seconds.",
17:18:50 /workspace/kubernetes/pkg is out of date. Please run hack/update-generated-swagger-docs.sh
17:18:50 !!! Error in ./hack/../hack/verify-generated-swagger-docs.sh:28
17:18:50   '"${KUBE_ROOT}/hack/after-build/verify-generated-swagger-docs.sh" "$@"' exited with status 1
17:18:50 Call stack:
17:18:50   1: ./hack/../hack/verify-generated-swagger-docs.sh:28 main(...)
17:18:50 Exiting with status 1
17:18:50 FAILED

**Comment by timstclair** [comment]
Ran `hack/update-all.sh` & squashed.

**Comment by timstclair** [comment]
@k8s-bot test this

**Comment by timstclair** [comment]
Looks like there's a problem with `hack/verify-codecgen.sh`, which says I need to rerun `hack/update-codecgen.sh`, but when I do it doesn't make any changes.

**Comment by timstclair** [comment]
@wojtek-t

**Comment by wojtek-t** [comment]
>
Looks like there's a problem with hack/verify-codecgen.sh, which says I need to rerun hack/update-codecgen.sh, but when I do it doesn't make any changes.

That's strange - both scripts are generating files using exactly the same command...

**Comment by timstclair** [comment]
Here's the output: https://gist.github.com/timstclair/cb26775ed0a89e6467cb

The code that's getting replaced looks weird. E.g. `if false {` and empty else if blocks.

**Comment by timstclair** [comment]
@k8s-bot test this

File: pkg/api/v1/types.go [hunk 2]
```diff
@@ -846,17 +846,27 @@ type ExecAction struct {
...
 	// More info: http://releases.k8s.io/HEAD/docs/user-guide/pod-states.md#container-probes
 	InitialDelaySeconds int64 `json:"initialDelaySeconds,omitempty"`
-	// Number of seconds after which liveness probes timeout.
+	// Number of seconds after which the probe times out.
 	// Defaults to 1 second.
 	// More info: http://releases.k8s.io/HEAD/docs/user-guide/pod-states.md#container-probes
 	TimeoutSeconds int64 `json:"timeoutSeconds,omitempty"`
+	// How often (in seconds) to perform the probe.
+	// Default to 10 seconds.
+	PeriodSeconds int `json:"periodSeconds,omitempty"`
```
Comment (bgrant0607) [inline pkg/api/v1/types.go]: Minimum value is 1?

File: pkg/api/v1/types.go [hunk 3]
```diff
@@ -846,17 +846,27 @@ type ExecAction struct {
...
+	// Minimum consecutive successes for the probe to be considered successful after having failed.
+	// Defaults to 1. Must be 1 for liveness.
+	SuccessThreshold int `json:"successThreshold,omitempty"`
```
Comment (bgrant0607) [inline pkg/api/v1/types.go]: Minimum value is 1?

File: pkg/api/v1/types.go [hunk 4]
```diff
@@ -846,17 +846,27 @@ type ExecAction struct {
...
+	// Minimum consecutive failures for the probe to be considered failed after having succeeded.
+	// Defaults to 3.
+	FailureThreshold int `json:"failureThreshold,omitempty"`
```
Comment (bgrant0607) [inline pkg/api/v1/types.go]: Minimum value is 1?

File: pkg/api/v1/types.go [hunk 5]
```diff
@@ -846,17 +846,27 @@ type ExecAction struct {
...
+// Probe describes a health check to be performed against a container to determine whether it is
+// alive or ready to recieve traffic.
 type Probe struct {
 	// The action taken to determine the health of a container
 	Handler `json:",inline"`
 	// Number of seconds after the container has started before liveness probes are initiated.
 	// More info: http://releases.k8s.io/HEAD/docs/user-guide/pod-states.md#container-probes
 	InitialDelaySeconds int64 `json:"initialDelaySeconds,omitempty"`
-	// Number of seconds after which liveness probes timeout.
+	// Number of seconds after which the probe times out.
```
Comment (bgrant0607) [inline pkg/api/v1/types.go]: Minimum value is 1?

File: pkg/api/v1/types.go (hunk 2 above)
Comment (bgrant0607) [inline pkg/api/v1/types.go]: Why is this int where the other Seconds parameters are int64?

**Comment by piosz** [comment]
@timstclair could you please rebase?

**Comment by timstclair** [comment]
Rebased & addressed @bgrant0607 's comments.

File: pkg/api/v1/types.go (hunk 5 above)
Comment (timstclair) [inline pkg/api/v1/types.go]: Done.

File: pkg/api/v1/types.go (hunk 4 above)
Comment (timstclair) [inline pkg/api/v1/types.go]: Done.

File: pkg/api/v1/types.go (hunk 3 above)
Comment (timstclair) [inline pkg/api/v1/types.go]: Done.

File: pkg/api/v1/types.go (hunk 2 above)
Comment (timstclair) [inline pkg/api/v1/types.go]: Done. No reason for this, changed to int64 for consistency (although I suspect we'll never need a period over 68 years).

**Comment by dchen1107** [comment]
LGTM

**Comment by timstclair** [comment]
Lowered `FailureThreshold` from default (3) for some restart tests to decrease flakiness (and fail more reliably in some cases).

**Comment by timstclair** [comment]
@k8s-bot test this

**Comment by dchen1107** [comment]
LGTM

**Comment by wojtek-t** [comment]
@timstclair - please rebase

**Comment by timstclair** [comment]
Rebased.

File: api/swagger-spec/v1.json [hunk 6]
```diff
@@ -13524,7 +13524,7 @@
    },
    "v1.Probe": {
     "id": "v1.Probe",
-    "description": "Probe describes a liveness probe to be examined to the container.",
+    "description": "Probe describes a health check to be performed against a container to determine whether it is alive or ready to recieve traffic.",
```
Comment (bgrant0607) [inline api/swagger-spec/v1.json]: Typo: receive

Comment (timstclair) [inline api/swagger-spec/v1.json]: https://github.com/kubernetes/kubernetes/pull/17550

**Comment by therc** [comment]
These don't seem to be documented anywhere other than in the API reference...

**Pull request merged**